### ISSUE TITLE
False positives for [no-unused-prop-types] when props is assigned to a local variable

### ISSUE THREAD

**Issue by laumair** [yannickcr/eslint-plugin-react#2177]
```
class App extends Component {
  static propTypes = {
    notifications: PropTypes.array.isRequired
  };

  customizeNotifications() {
    const props = this.props;

    return props.notifications.map((notification) => `foo${notification}`);
  }

  render() {
    const notifications = this.customizeNotifications();

    return (
      <View>
        {notifications.map((notification) => <Text>{notification}</Text>)}
      </View>
    );
  }
```

Eslint complains about `notifications` prop as unused.

**Comment by ljharb** [comment]
Presumably if you `const { notifications } = this.props` instead, it passes?

**Comment by laumair** [comment]
> Presumably if you `const { notifications } = this.props` instead, it passes?

Yep, it passes on destructuring props.

**Comment by ljharb** [comment]
Although I’d strongly recommend that pattern, yours should certainly work if possible.

**Comment by jzabala** [comment]
@ljharb the optimal solution would be to handle every possible edge case but things can get out of hand very quickly. Like:

- Passing the props as argument `renderSomething(this.props)`
- Or reassigning props more that one:
```js
var props = this.props
var moreProps = {more: true, ...props}
// ... do something with moreProps
```

In the case presented by this issue what would be the added value of this:
```js
customizeNotifications() {
    const props = this.props;
    return props.notifications.map((notification) => `foo${notification}`);
  }
```
vs
```
customizeNotifications() {
    return this.props.notifications.map((notification) => `foo${notification}`);
  }
```

**Comment by ljharb** [comment]
I don't think there's much added value - they should do `const { notifications } = this.props;` regardless :-) but it's nice when we can correctly analyze as much as possible.

I think eslint provides a number of utilities to make tracking the source of a variable easier?

**Comment by jzabala** [comment]
> I don't think there's much added value - they should do `const { notifications } = this.props;` regardless :-) but it's nice when we can correctly analyze as much as possible.

Awesome.

> I think eslint provides a number of utilities to make tracking the source of a variable easier?

Yeah, you can use `context.getDeclaredVariables` for that. In this case though is wasn't necessary since it seems it was fixed by another PR. I created a PR with a test for this issue.
